### ISSUE TITLE
There is no easy way to create an isolated dotnet installation

### ISSUE THREAD

**Issue by noahfalk** [dotnet/core-setup#3361]
## Steps to reproduce

Install a copy of dotnet 2.0.0 to a non-default location, such as C:\private\.dotnet using the DotNet-Install.ps1 
Put the private copy of dotnet on the path
Run a command such as dotnet store

## Expected  behavior

The dotnet app loads all dotnet binaries from within C:\private\.dotnet\

## Actual behavior

Some binaries are loaded from global locations, such as C:\Program Files\dotnet\sdk\15.5.0-preview-007044\dotnet.dll

## Environment data

dotnet --info
.NET Command Line Tools (2.0.0)

Product Information:
 Version:            2.0.0
 Commit SHA-1 hash:  cdcd1928c9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   F:\github\JitBench\.dotnet\sdk\2.0.0\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d

## Comments

This behavior is by design as far as I can tell, however I am proposing this design is not ideal and we should investigate changing it in the scope of this scenario at least. The current design has several downsides:

- It isn't the behavior users expected. The commonly held belief among devs I've spoken with is that creating a copy of the dotnet directory at a private location and executing dotnet.exe there has given them an isolated installation that loads dependencies from the same folder.
- Development and testing is more error-prone. The current design makes it very easy to accidentally load a binary from the global location instead of the one you thought you were testing.
- Repro environments are much harder to capture and reproduce accurately. For example see the environment data in this bug. If that is all the info we knew and I hadn't separately debugged/explicitly called out that 15.5.0-preview is actually the dotnet that loads its very likely any bug I hit in that configuration would have been resolved 'no repro'.

## Workarounds
I am currently using the environment var DOTNET_MULTILEVEL_LOOKUP=0 as a workaround. It is better than nothing, but I don't think it is a high quality solution on its own:

- Developers don't know about it, and because they don't expect this behavior in the first place they won't think to look for it.
- Environment variables do not persist for the lifetime of an installation. It is easy to forget to set it when running many commands over multiple sessions or switching between shells.
- It is incomplete (https://github.com/dotnet/core-setup/issues/3360)

**Comment by Petermarcu** [comment]
@noahfalk, is this just happening for lookups in the sdk or for runtime bits as well?

@livarcocc

**Comment by noahfalk** [comment]
I only observed it live for a few SDK binaries, by code review it appears to happen for store binaries as well (#3360), and I haven't reviewed the code or observed evidence of it for runtime IL binaries.

**Comment by steveharter** [comment]
Note that #3360 is fixed for 2.1 (store binaries loaded when DOTNET_MULTILEVEL_LOOPUP=0); it is being considered to port to 2.0.x.

The SDK binaries will be loaded when doing a `--info` and will be loaded from the newest version of the SDK whether local or global (by design).

**Comment by steveharter** [comment]
@noahfalk I re-tested this on 2.1.0 with DOTNET_MULTILEVEL_LOOPUP=0 and the local SDK was selected (even though there was a newer global version) and no assemblies were loaded from the global locations.

I do think that it is possible to only use the location of dotnet.exe and no longer search in the global location. We already do this on unix. However, we cannot change this (e.g. don't use multi-level lookup on Windows) for a minor release as it would be breaking, so it would need to wait until 3.0.

Moving this to future.

**Comment by danmosemsft** [comment]
Would it make sense in 2.1 to expose a flag on dotnet.exe equivalent to DOTNET_MULTILEVEL_LOOKUP=0? That would be more discoverable (and memorable - I have to refer to notes to remember the name)

**Comment by steveharter** [comment]
Yes a config option or command line option is possible provided it is does not disable multilevel by default. However, I think the intent of the issue was to change the default:
>> Developers don't know about it, and because they don't expect this behavior in the first place they won't think to look for it.

**Comment by danmosemsft** [comment]
Is there a conclusion that we should change the default? I had the same experience as @noahfalk above, and was surprised, and had to be told about the variable. However @eerhardt explained this was done because it means you get patches, and the version numbers should ensure it is safe.
(Clearly the version numbers aren't very useful in this context for developers like us working on CoreFX who eg want to test an old version)

**Comment by steveharter** [comment]
>> Is there a conclusion that we should change the default

I don't think there is a consensus that we should change the default; we haven't discussed this amongst the stakeholders yet.

There is the development vs production scenario. For development, yes it is annoying we look in the global location by default when someone is trying to use a specific local version (not exactly falling into the "pit of success"). However for production I think they would favor the multilevel global location semantics for servicing, and I don't know why they would install in a local folder anyway.

I do think the concept of a well-known global location is useful. Consider if we add support for <appname>.exe that uses the shared framework (not standalone). This has been requested. We would need to look in a well-known location because we don't run dotnet.exe and thus can't use its location. Also it is somewhat inconsistent that there isn't a way to add a global location in unix, but that's because unix doesn't have a concept of a shared program location like Windows, although we could add an environment variable to specify the global location.

Also note that we have the "hammer" servicing logic which will look in a global location for critical updates.

Anyway, I think a config option and\or a command-line option for 2.1.0 would be a reasonable request.

**Comment by noahfalk** [comment]
I can certainly be sympathetic to production desires for servicing, though even there I think we may need to do more to help people get the behavior they individually want? I could imagine there are some people in each of these groups:
a) I want my app to have total isolation, if there is any need for servicing whatsoever I (as the dev) will handle it
b) I want my app to receive security updates, but otherwise leave it alone
c) I want my app to receive any updates that are expected to be reasonably back compatible (perf improvements, new lightup features, diagnostics improvements). I shipped my app with a standalone copy of .Net Core to ensure it still works even when no global install is present on the machine, not to avoid updates.

I assume to have a good discussion about changing the default we'd need to gather data about what most of our customers want if we don't already have that feedback. 

If we could have a config file that would certainly be very helpful in getting the behavior each person wants. It allows us to have scripts like Dotnet-Install.ps1 set a config that avoids updates. This feels like it would satisfy the "developer of .Net Core" scenario without forcing the production scenarios to agree on the same policy.
